src/azure-cli/azure/cli/command_modules/synapse/operations/accesscontrol.py

```diff
@@ -4,55 +4,67 @@
 # --------------------------------------------------------------------------------------------
 
 from knack.util import CLIError
-from azure.synapse.accesscontrol.models import RoleAssignmentOptions
 from azure.cli.core.util import is_guid
 from azure.graphrbac.models import GraphErrorException
 from msrestazure.azure_exceptions import CloudError
-from .._client_factory import cf_synapse_client_accesscontrol_factory, cf_graph_client_factory
+from .._client_factory import cf_synapse_role_assignments, cf_synapse_role_definitions, cf_graph_client_factory
+from ..constant import ITEM_NAME_MAPPING
 
 
 # List Synapse Role Assignment
-def list_role_assignments(cmd, workspace_name, role=None, assignee=None):
-    # get role id
-    role_id = _resolve_role_id(cmd, role, workspace_name)
-    # get object_id
-    object_id = _resolve_object_id(cmd, assignee, fallback_to_object_id=True)
+def list_role_assignments(cmd, workspace_name, role=None, assignee=None, assignee_object_id=None,
+                          scope=None, item=None, item_type=None):
+    if bool(assignee) and bool(assignee_object_id):
+        raise CLIError('usage error: --assignee STRING | --assignee-object-id GUID')
+
+    if bool(item) != bool(item_type):
+        raise CLIError('usage error: --item-type STRING --item STRING')
+
+    return _list_role_assignments(cmd, workspace_name, role, assignee or assignee_object_id,
+                                  scope, resolve_assignee=(not assignee_object_id), item=item, item_type=item_type)
 
-    client = cf_synapse_client_accesscontrol_factory(cmd.cli_ctx, workspace_name)
-    role_assignments = client.get_role_assignments(role_id, object_id)
 
-    # TODO:
-    # Currently, when only `ObjectId` is specified, the cmdlet returns incorrect result.
-    # Filter from client side as a workaround
-    if object_id:
-        role_assignments = [x for x in role_assignments if x.principal_id == object_id]
+def _list_role_assignments(cmd, workspace_name, role=None, assignee=None, scope=None,
+                           resolve_assignee=True, item=None, item_type=None):
+    """Prepare scope, role ID and resolve object ID from Graph API."""
+    if any([scope, item, item_type]):
+        scope = _build_role_scope(workspace_name, scope, item, item_type)
+    role_id = _resolve_role_id(cmd, role, workspace_name)
+    object_id = _resolve_object_id(cmd, assignee, fallback_to_object_id=True) if resolve_assignee else assignee
+    client = cf_synapse_role_assignments(cmd.cli_ctx, workspace_name)
+    role_assignments = client.list_role_assignments(role_id, object_id, scope).value
+    if scope:
+        role_assignments = [x for x in role_assignments if x.scope == scope]
```

Comment:
Why do we need to filter the results by scope again since `client.list_role_assignments` already take `scope` parameter?

Reply:
Oh I saw the previous comments and got the reason. Better to keep the comments~

Reply:
Good catch. I think we should remove the logic, just return raw RP response. @sunsw1994

Reply:
The server side has not supported list assignments by scope. `scope` parameter will be ignored by server, so I add filter at CLI side.